File: internal/hir/ids.go
package hir

type ModuleID uint32
type FunctionID uint32
type BlockID uint32
type ValueID uint32
type TypeID uint32
type ShapeID uint32

func NewModuleID(value uint32) ModuleID     { return ModuleID(value) }
func NewFunctionID(value uint32) FunctionID { return FunctionID(value) }
func NewBlockID(value uint32) BlockID       { return BlockID(value) }
func NewValueID(value uint32) ValueID       { return ValueID(value) }
func NewTypeID(value uint32) TypeID         { return TypeID(value) }
func NewShapeID(value uint32) ShapeID       { return ShapeID(value) }

File: crates/tsnative-ir/src/ids.rs
use serde::{Deserialize, Serialize};

macro_rules! id_type {
    ($name:ident) => {
        #[derive(
            Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash, Serialize, Deserialize,
        )]
        pub struct $name(pub u32);

        impl $name {
            pub const fn new(value: u32) -> Self {
                Self(value)
            }
        }
    };
}

id_type!(ModuleId);
id_type!(FunctionId);
id_type!(BlockId);
id_type!(ValueId);
id_type!(TypeId);
id_type!(ShapeId);
